Allow "Difficulty Adjust" mod's extended AR selection to go below zero

isakvik opened the pull request:
Discussion reference: #22898

This PR adds negative ARs down to -10 to the difficulty adjust mod for osu!standard (for the extended range), and adds millisecond value information to the selection slider.

Negative ARs are fun, and well supported in the gamemode, so this simple implementation has been made to enable eventual sanity checking of even lower ARs, or with a more complicated UI solution if the slider is deemed insufficient.

![Adjustment screen](https://gn.s-ul.eu/jp4AzMOW)

isakvik committed:
extendable minimum AR value for osu!std diff adjust mod, set to -10

isakvik committed:
adds new diffsetting control that shows millisecond values for approach rate override

peppy committed:
Fix code quality issues

peppy committed:
Inline local control

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModDifficultyAdjust.cs:
```diff
@@ -53,5 +58,31 @@ protected override void ApplySettings(BeatmapDifficulty difficulty)
+
+                private double millisecondsFromApproachRate(float value, float clockRate)
+                {
+                    return Math.Round(1800 - Math.Min(value, 5) * 120 - (value >= 5 ? (value - 5) * 150 : 0) / clockRate);
```
I don't know where you got this from, but it doesn't match any existing implementations and is probably wrong. I'll fix it.

bdach replied:
Fixed by just creating a hitcircle and reading the raw value out of it.

I also nuked the `clockRate` param (I get what this was going for, but we're not handling "effective" AR anywhere yet, so this should be done separately and comprehensively), and made the ms thing display only when extended limits are active. That last one is kinda style, kinda not wanting to overwhelm users / prompt questions "why does only osu! ruleset get this ms display".

bdach committed:
Fix too many things being exposed

bdach committed:
Remove local preempt time calculation

And just use a hitcircle, and read the actual value. Comes with 100%
less chance of forgetting to update either place in the future.

bdach committed:
Only show preempt time when limits are extended

apollo-dw commented:
I just want to quickly step in and say we could consider adjusting the AR -> visibility time formula as suggested in the discussion. Above AR5 and below AR5 have different formulas IIRC, so there is precedent for adjusting the formula for below 0.

What this would do in effect is allow for way longer visibility times, *without* making the scaling of the slider super weird. The millisecond information appearing in the tooltip only makes this more feasible IMO. I've seen reading players do ARs lower than even -20...

peppy commented:
> I just want to quickly step in and say we could consider adjusting the AR -> visibility time formula as suggested in the discussion. Above AR5 and below AR5 have different formulas IIRC, so there is precedent for adjusting the formula for below 0.

This becomes a performance and UX issue. Let's track that separately.

I'm only pushing this one through so we can handle the seemingly major case and cease the discussions as to keep focus on the more important tasks at hand.
